File: src/main/python/movie_lens_ranker/RecommendedMovies.py
from typing import Tuple

import numpy as np
from array_record.python import array_record_module
import msgpack

class RecommendedMovies (object):
    def __init__(self, num_users:int, movie_rec_file_uri:str, movie_rec_ts_file_uri:str):
        """
        read in the recommended movies.  assumes that each list of recommended movies has the same length, and
        that a user_id netry in movie_rec_file_path is complemented by a user_id entry in
        movie_rec_ts_file_path.
        :param num_users:  number of users in the entire user catalog
        :param movie_rec_file_uri: file of movies recommended for each user in format such that each row is
        usr_id, [movie_ids].
        :param movie_rec_ts_file_uri: file of timestamps of movies recommended for each user in format such that each row is
        user_id, [timestamps] and the row for a user_id in this file is complementary to the row of same user_id in
        movie_rec_file_path and both lists of movie_ids and timestamps are ordered to represent the same movies.
        notr that recommended movies which have not been seen should have a timestamp larger than feasibly in the
        watch history, like ts_2050 = 2524608000 used by the retrieval project.
        """
        self.pad_value = -1
        self.user_ids, self.movies, self.row_length = self._read(num_users, file_uri=movie_rec_file_uri, pad_value=self.pad_value)
        _ , self.timestamps, _ = self._read(num_users, movie_rec_ts_file_uri,  pad_value=self.pad_value, dtype2=np.int64)
        self.num_users = num_users
        
        
    def _read(self, num_users:int, file_uri:str, pad_value:int, dtype2:type=np.int32) -> Tuple[np.ndarray, np.ndarray, int]:
        """
        read the recommended movies or recommended movies' timestamps file.   note that the array_records
        have been written such that the movies decrease in score and such that the
        timestamps are for the movie_ids at same index in orther file.
        :param num_users: total number of users in the user catalog.  this is probably the same as
        the length of the file at file_uri
        :param file_uri:
        :param dtype2:
        :return:
        """
        reader = None
        try:
            reader = array_record_module.ArrayRecordReader(file_uri)
            
            # peek at length of record[1]
            row0_list = [msgpack.unpackb(b, use_list=False) for b in reader.read([0])]
            len_arr = len(row0_list[0][1])
            
            user_ids = np.array([i for i in range(0, num_users + 1)], dtype=np.int32)
            #the padding value doesn't matter because items gets completely filled with real values
            items = np.full((num_users + 1, len_arr), fill_value=pad_value, dtype=dtype2)
            
            count = reader.num_records()
            batch_bytes = reader.read([x for x in range(0, count)])
            records = [msgpack.unpackb(b, use_list=False) for b in batch_bytes]
            for record in records:
                user_idx = record[0]
                items[user_idx, :] = record[1]
        finally:
            if reader is not None:
                reader.close()
        return user_ids, items, len_arr
    
    def get_unseen_movies(self, user_id: np.ndarray, timestamp: np.ndarray, top_k:int=200) -> np.ndarray:
        """
        given array of user_ids, return top_k recommended movies for user that they haven't seen before time=timestamp.
        the unseen movies have been moved to front of array and retain their respective original order which is by decreasing similarity score.
        Note that if top_k is > the file's
        :param user_id: input array of shape (None,), e.g. np.array([2,4])
        :param timestamp: timestamp representing current time.  any recommendations with timestamps > timestamp are yet unseen.
        :param top_k: number of top unseen recommendations to return
        :return: top k of movie recommendations unseen by user_id.  shape returned is (len(user_id_, top_k)
        """
        if top_k > self.row_length:
            raise ValueError(f"top_k must be smaller than the number of recommendations per suer = {self.row_length}")
        
        user_idx = self.user_ids[user_id]
        
        mask = self.timestamps[user_idx] > timestamp[:, np.newaxis]
        sort_indices = np.argsort(~mask, axis=1, kind='stable')
        #equiv of row-wise tf.gather:
        sel_movies = np.take_along_axis(self.movies[user_idx], sort_indices, axis=1)
        
        #the unseen movies have been moved to front of array and retain their respective original order which is by decreasing similarity score.
        return sel_movies[:, :top_k]

File: src/main/rust/src/recommended_movies.rs

use arrow_array::{Int32Array, Int64Array, FixedSizeListArray};
use tokio::runtime::Runtime;
use object_store::ObjectStoreExt;
use futures::stream::StreamExt;
use parquet::arrow::async_reader::ParquetObjectReader;
use parquet::arrow::async_reader::ParquetRecordBatchStreamBuilder;
use arrow_schema::DataType;
use crate::util;

//NOTE: unlike the python version, the rust version uses implicit user indices for the movie_ids
// and timestamp rows that are 1 through num_users.  the python datastructures use 0 through num_users - 1
// movie_ids and timestamps are shape  (num_users * num_movies)
pub struct RecommendedMovies {
    pub movie_ids : Vec<i32>,
    pub timestamps : Vec<i64>,
    num_users: usize,
    pad_value : i32,
    ts_pad_value : i64,
}

impl RecommendedMovies {
    /*
    given array of user_ids, return top_k recommended movies for user that they haven't seen before time=timestamp.
    the unseen movies have been moved to front of array and retain their respective original order which is by decreasing similarity score.

    Arguments:
    * user_id: input array of shape (None,), e.g. np.array([2,4])
    * timestamp: timestamp representing current time.  any recommendations with timestamps > timestamp are yet unseen.
    * top_k: number of top unseen recommendations to return

    Returns:
     top k of movie recommendations unseen by user_id.  shape returned is (len(user_id_, top_k)

     */
    pub fn get_unseen_movies(
        &self,
        user_ids: Vec<i32>,
        timestamps: Vec<i64>,
        top_k: usize
    ) -> Vec<i32> {
        // Derive num_movies dynamically from the flat array length
        let num_movies = self.movie_ids.len() / (self.num_users);

        if top_k > num_movies {
            // Equivalent to raising ValueError
            panic!("top_k ({}) must be smaller than the number of recommendations per user ({})", top_k, num_movies);
        }

        let n_requests = user_ids.len();
        let mut ret_movie_ids = vec![self.pad_value; n_requests * top_k];

        for (i, (&target_uid, &target_ts)) in user_ids.iter().zip(timestamps.iter()).enumerate() {

            // Assuming 1-based user_ids mapping to 0-based indices based on your previous code
            let user_idx = (target_uid - 1) as usize;

            // Locate the user's data chunk in the flattened 1D array
            let start_idx = user_idx * num_movies;
            let end_idx = start_idx + num_movies;

            let user_movies = &self.movie_ids[start_idx..end_idx];
            let user_timestamps = &self.timestamps[start_idx..end_idx];

            let out_offset = i * top_k;
            let mut count = 0;

            // PASS 1: Collect "Unseen" movies first (timestamps > target_ts)
            // This is the equivalent of argsort putting ~mask (0s) at the front
            for j in 0..num_movies {
                if count >= top_k { break; }

                if user_timestamps[j] > target_ts {
                    ret_movie_ids[out_offset + count] = user_movies[j];
                    count += 1;
                }
            }

            // PASS 2: If we haven't reached top_k yet, fill the rest with "Seen" movies
            // This is the equivalent of argsort putting ~mask (1s) at the back, preserving their order
            if count < top_k {
                for j in 0..num_movies {
                    if count >= top_k { break; }

                    if user_timestamps[j] <= target_ts {
                        ret_movie_ids[out_offset + count] = user_movies[j];
                        count += 1;
                    }
                }
            }
        }

        ret_movie_ids
    }
}


/// Efficiently reads complementary movies and timestamps files into flattened 2D arrays.
/// Returns a tuple of (movie_ids_flat, timestamps_flat) each matching the shape (num_users, num_movies)
pub async fn read_recommendation_files(
    movies_uri: &str,
    timestamps_uri: &str,
    num_users: usize,
    num_movies: usize,
    pad_value: i32,
    ts_pad_value : i64
) -> (Vec<i32>, Vec<i64>) {

    let movies_uri = movies_uri.to_string();
    let timestamps_uri = timestamps_uri.to_string();

    let movies_task = tokio::spawn(async move {
        let mut movies_data = vec![pad_value; num_users * num_movies]; // Pre-allocate with a pad value

        let (storage, path) = util::parse_uri(&movies_uri);
        let meta = storage.head(&path).await.expect("Failed to get movies metadata");
        let reader = ParquetObjectReader::new(storage, meta.location).with_file_size(meta.size);
        let builder = ParquetRecordBatchStreamBuilder::new(reader).await.unwrap();
        let mut stream = builder.build().unwrap();

        while let Some(batch_result) = stream.next().await {
            let batch = batch_result.expect("Failed to read movies RecordBatch");

            let user_ids = batch.column(0).as_any().downcast_ref::<Int32Array>().unwrap();
            // Downcast the nested List column
            let movie_lists = batch.column(1).as_any().downcast_ref::<FixedSizeListArray>().unwrap();

            for i in 0..batch.num_rows() {
                let uid = user_ids.value(i);
                // Map 1-based user_id directly to a 0-based array index
                let user_idx = (uid - 1) as usize;
                let offset = user_idx * num_movies;

                // Extract the inner primitive array from the List row
                let list_element = movie_lists.value(i);
                let movie_values = list_element.as_any().downcast_ref::<Int32Array>().unwrap();

                // Blast the nested array data directly into our pre-allocated matrix slice via memcpy
                movies_data[offset..offset + num_movies].copy_from_slice(movie_values.values());
            }
        }
        movies_data
    });

    // Task 2: Process the Timestamps file concurrently
    let timestamps_task = tokio::spawn(async move {
        let mut timestamps_data = vec![ts_pad_value; num_users * num_movies];

        let (storage, path) = util::parse_uri(&timestamps_uri);
        let meta = storage.head(&path).await.expect("Failed to get timestamps metadata");
        let reader = ParquetObjectReader::new(storage, meta.location).with_file_size(meta.size);
        let builder = ParquetRecordBatchStreamBuilder::new(reader).await.unwrap();
        let mut stream = builder.build().unwrap();

        while let Some(batch_result) = stream.next().await {
            let batch = batch_result.expect("Failed to read timestamps RecordBatch");

            let user_ids = batch.column(0).as_any().downcast_ref::<Int32Array>().unwrap();
            let ts_lists = batch.column(1).as_any().downcast_ref::<FixedSizeListArray>().unwrap();

            for i in 0..batch.num_rows() {
                let uid = user_ids.value(i);
                let user_idx = (uid - 1) as usize;
                let offset = user_idx * num_movies;

                let list_element = ts_lists.value(i);
                let ts_values = list_element.as_any().downcast_ref::<Int64Array>().unwrap();

                timestamps_data[offset..offset + num_movies].copy_from_slice(ts_values.values());
            }
        }
        timestamps_data
    });

    // Await both tasks together. They run completely in parallel!
    let (movies_res, timestamps_res) = tokio::join!(movies_task, timestamps_task);

    (
        movies_res.expect("Movies processing panicked"),
        timestamps_res.expect("Timestamps processing panicked"),
    )
}

fn read_num_movies(movies_uri: &String) -> usize {
    let rt = Runtime::new().unwrap();

    rt.block_on(async {
        let (storage, path) = util::parse_uri(movies_uri);
        let meta = storage.head(&path).await.expect("Failed to get movies metadata");
        let reader = ParquetObjectReader::new(storage, meta.location).with_file_size(meta.size);
        let builder = ParquetRecordBatchStreamBuilder::new(reader).await.unwrap();

        let schema = builder.schema();

        // 3. Fetch the field by name and inspect its Arrow DataType
        match schema.field_with_name("movie_ids").unwrap().data_type() {
            DataType::FixedSizeList(_inner_field, size) => *size as usize,
            _ => panic!("Expected movie_ids to be a FixedSizeList schema! Check your PyArrow writer."),
        }
    })
}

/// build the RecommendedMovies from the given list of uris for the movie recommendation parquet files
/// # Arguments
/// * `ratings_uris` -  vector of uris for ratings parquet files for building UserHistory hashmap
pub fn build_recommended_movies(num_users:usize, movie_rec_file_uri:&String, movie_rec_ts_file_uri:&String) -> RecommendedMovies {

    let pad_value : i32 = -1;
    let ts_pad_value: i64 = 2524608000; //year 2050

    let num_movies = read_num_movies(&movie_rec_file_uri);

    // Because PyO3 functions are synchronous by default, we spin up a
    // Tokio runtime to execute our highly concurrent async disk/network I/O.
    // The code can begin processing data while it is still loading because the loading onto CPU is usually a bottleneck.
    let rt = Runtime::new().unwrap();

    // Block the thread until the map is built and sorted
    let (movie_ids, timestamps)  = rt.block_on(read_recommendation_files(
        &movie_rec_file_uri, &movie_rec_ts_file_uri, num_users, num_movies, pad_value, ts_pad_value));

    RecommendedMovies { movie_ids, timestamps, num_users, pad_value, ts_pad_value}

}